### PR TITLE
Fix various bogus unreachable code warnings

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes the re-opened <https://bugs.swift.org/browse/SR-2729>. Also might fix <https://bugs.swift.org/browse/SR-3696> but I don't have a Linux machine available right now.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@jckarter what's the difference between CleanupLocation and calling SILLocation::markAutoGenerated()?

File: lib/SIL/SILType.cpp [hunk 1]
```diff
@@ -638,17 +638,17 @@ bool SILModuleConventions::isPassedIndirectlyInSIL(SILType type, SILModule &M) {
 }
 
 bool SILFunctionType::isNoReturnFunction() {
-  return getDirectFormalResultsType().getSwiftRValueType()->isUninhabited();
+  if (getNumResults() == 1)
+    return getSingleResult().getType()->isUninhabited();
```
Comment (jckarter) [inline lib/SIL/SILType.cpp]: A function would be noreturn if *any* of its results are uninhabited, wouldn't it?

**Comment by jckarter** [comment]
AIUI both have the same ultimate effect of being treated as implicit by the debugger.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**